Add operator<=> to CompositionHighlight and CompositionUnderline for natural ordering

stwrt opened the pull request:
#### 443314896cd23eb23412a812145d1becd8a2d542
<pre>
Add operator&lt;=&gt; to CompositionHighlight and CompositionUnderline for natural ordering
<a href="https://bugs.webkit.org/show_bug.cgi?id=303195">https://bugs.webkit.org/show_bug.cgi?id=303195</a>
<a href="https://rdar.apple.com/problem/165485853">rdar://problem/165485853</a>

Reviewed by Sam Weinig.

Replace lambdas in std::ranges::sort calls with operator&lt;=&gt; definitions on CompositionHighlight and CompositionUnderline.
This removes the multi-line comparison logic in the lambdas.

* Source/WebCore/editing/CompositionHighlight.h:
(WebCore::CompositionHighlight::operator&lt;=&gt;):
* Source/WebCore/editing/CompositionUnderline.h:
(WebCore::CompositionUnderline::operator&lt;=&gt;):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(extractUnderlines):
(compositionHighlights):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::compositionHighlights):
(WebKit::compositionUnderlines):

Canonical link: <a href="https://commits.webkit.org/303640@main">https://commits.webkit.org/303640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f503b5a76603c6ae861ab0139812f9bde0f3255c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140663 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85157 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/09fa50a5-e523-4fdd-aea8-5afca429ea17) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134988 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6117 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5482 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101806 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69183 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/997f6ec9-54ad-469c-acd5-a4dadbdf6ebc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136065 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4319 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119302 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82603 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/daf1776c-7dbf-4203-bf4d-534a4939fc3e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4203 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1784 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143309 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5289 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37997 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110185 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5371 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110365 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27969 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4080 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115559 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58945 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5344 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33913 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5187 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68796 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5433 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5300 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5574,12 +5574,8 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
-        if (a.startOffset > b.startOffset)
-            return false;
-        return a.endOffset < b.endOffset;
+    std::ranges::sort(highlights, { }, [](auto& highlight) {
```
Another direction to go here, that might be a bit nicer long term, would be to add operator<=> implementations for WebCore::CompositionHighlight and WebCore::CompositionUnderline in WebCore. Then you wouldn't need the project at all.

stwrt replied:
Good catch. That simplifies the sort logic even further, so we can just call `sort` without extra parameters.

weinig commented on Source/WebCore/editing/CompositionHighlight.h:
```diff
@@ -50,6 +50,13 @@ struct CompositionHighlight {
     std::optional<Color> backgroundColor;
     std::optional<Color> foregroundColor;
+
+    friend bool operator<(const CompositionHighlight& a, const CompositionHighlight& b)
```
Did you consider using the <=> operator? I think it could be implemented as:

```c++
std::strong_ordering operator<=>(const CompositionHighlight& other) const
{
    return std::make_pair(startOffset, endOffset) <=> std::make_pair(other.startOffset, other.endOffset);
}
```

stwrt replied:
Adopted this method again. Was able to get it to build with just adding the `==` operator too.

weinig commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6631,12 +6631,8 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
-        if (a.startOffset > b.startOffset)
-            return false;
-        return a.endOffset < b.endOffset;
+    std::ranges::sort(underlines, [](const auto& a, const auto& b) {
```
What was the compile error you ran into when using just `std::ranges::sort(underlines);`?

weinig reviewed (approved):
Nice!

stwrt committed:
Add operator<=> to CompositionHighlight and CompositionUnderline for natural ordering
https://bugs.webkit.org/show_bug.cgi?id=303195
rdar://problem/165485853

Reviewed by Sam Weinig.

Replace lambdas in std::ranges::sort calls with operator<=> definitions on CompositionHighlight and CompositionUnderline.
This removes the multi-line comparison logic in the lambdas.

* Source/WebCore/editing/CompositionHighlight.h:
(WebCore::CompositionHighlight::operator<=>):
* Source/WebCore/editing/CompositionUnderline.h:
(WebCore::CompositionUnderline::operator<=>):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(extractUnderlines):
(compositionHighlights):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::compositionHighlights):
(WebKit::compositionUnderlines):

Canonical link: https://commits.webkit.org/303640@main
